dlib/type_safe_union/type_safe_union_kernel.h

```diff
@@ -127,12 +118,16 @@ namespace dlib
             return detail::type_safe_union_type_id<T,Types...>::value;
         }
 
+        template <typename T>
+        static constexpr int get_type_id (in_place_tag<T>)
```

Comment:
Update the docs in the `_abstract.h` file since this is part of the public API.  Make sure all docs that need to be updated are updated. I'm still looking this over.  The PR looks nice though :)